supervision/detection/utils.py

```diff
@@ -691,23 +691,51 @@ def merge_data(
     if not data_list:
         return {}
 
-    all_keys_sets = [set(data.keys()) for data in data_list]
-    if not all(keys_set == all_keys_sets[0] for keys_set in all_keys_sets):
-        raise ValueError("All data dictionaries must have the same keys to merge.")
-
     for data in data_list:
-        lengths = [len(value) for value in data.values()]
-        if len(set(lengths)) > 1:
+        lengths_set = [len(value) for value in data.values()]
+        if len(set(lengths_set)) > 1:
             raise ValueError(
                 "All data values within a single object must have equal length."
             )
 
-    merged_data = {key: [] for key in all_keys_sets[0]}
-
+    all_keys: Set[str] = set()
     for data in data_list:
-        for key in merged_data:
-            merged_data[key].append(data[key])
+        all_keys.update(data.keys())
+
+    # Naively merging entries and then validating length comes with a problem:
+    # N values may come from data[0]["key_1"] and N values from data[1]["key_2"].
+    # These should not be joined together.
+    # Here, as soon as we find data of len > 0, we lock the key set and raise
+    # a ValueError if later we find a value of len > 0 with an unknown key.
+    key_set = None
+    merged_data = {key: [] for key in all_keys}
+    for data in data_list:
```

Comment:
Please limit the amount of indentions here. 
Two levels are okay, three is usually too many. An if inside an if, inside a for, inside another for is definitely too much. No wonder you had to add a comment explaining what's going on here. :)